pkg/ruler/manager.go

```diff
@@ -141,33 +137,50 @@ func (r *DefaultMultiTenantManager) syncRulesToManager(ctx context.Context, user
 		return
 	}
 
+	manager, done := r.getOrCreateManager(ctx, user, update)
+	if done {
+		return
+	}
+	err = manager.Update(r.cfg.EvaluationInterval, files, nil, r.cfg.ExternalURL.String(), nil)
+	if err != nil {
+		r.lastReloadSuccessful.WithLabelValues(user).Set(0)
+		level.Error(r.logger).Log("msg", "unable to update rule manager", "user", user, "err", err)
+		return
+	}
+
+	r.lastReloadSuccessful.WithLabelValues(user).Set(1)
+	r.lastReloadSuccessfulTimestamp.WithLabelValues(user).SetToCurrentTime()
+}
+
+// getOrCreateManager retrieves the user manager. if it doesn't exist, it will create and start it first
+func (r *DefaultMultiTenantManager) getOrCreateManager(ctx context.Context, user string, update bool) (RulesManager, bool) {
+	r.userManagerMtx.Lock()
+	defer r.userManagerMtx.Unlock()
+
 	manager, exists := r.userManagers[user]
-	if !exists || update {
-		level.Debug(r.logger).Log("msg", "updating rules", "user", user)
-		r.configUpdatesTotal.WithLabelValues(user).Inc()
-		if !exists {
-			level.Debug(r.logger).Log("msg", "creating rule manager for user", "user", user)
-			manager, err = r.newManager(ctx, user)
-			if err != nil {
-				r.lastReloadSuccessful.WithLabelValues(user).Set(0)
-				level.Error(r.logger).Log("msg", "unable to create rule manager", "user", user, "err", err)
-				return
-			}
-			// manager.Run() starts running the manager and blocks until Stop() is called.
-			// Hence run it as another goroutine.
-			go manager.Run()
-			r.userManagers[user] = manager
-		}
-		err = manager.Update(r.cfg.EvaluationInterval, files, nil, r.cfg.ExternalURL.String(), nil)
-		if err != nil {
-			r.lastReloadSuccessful.WithLabelValues(user).Set(0)
-			level.Error(r.logger).Log("msg", "unable to update rule manager", "user", user, "err", err)
-			return
-		}
+	if exists && !update {
+		return nil, true
+	}
+
+	level.Debug(r.logger).Log("msg", "updating rules", "user", user)
+	r.configUpdatesTotal.WithLabelValues(user).Inc()
```

Comment:
We're not doing any real update here, so I don't fully understand why you're passing `update` to `getOrCreateManager()`. I think all update related logic should be in the caller of `getOrCreateManager()`.